server/implementation/src/main/java/io/smallrye/graphql/spi/DataFetcherService.java

```diff
@@ -0,0 +1,147 @@
+        private EventEmitter eventEmitter;
+        private BatchLoaderHelper batchLoaderHelper;
+
+        public DataFetcherService newInstance() {
+            return new DefaultDataFetcherService();
+        }
+
+        @Override
+        public String getName() {
+            return "Reflection (default)";
+        }
+
+        @Override
+        public String[] forClasses() {
+            return null; // This is the default, so all non-specified classes
+        }
+
+        @Override
+        public void init(Operation operation, Config config) {
+            this.eventEmitter = EventEmitter.getInstance(config);
+            this.fieldHelper = new FieldHelper(operation);
+            this.reflectionHelper = new ReflectionHelper(operation, eventEmitter);
+            this.argumentHelper = new ArgumentHelper(operation.getArguments());
+            this.partialResultHelper = new PartialResultHelper();
+            this.batchLoaderHelper = new BatchLoaderHelper();
+        }
+
+        @Override
+        public <T> T get(DataFetchingEnvironment dfe, DataFetcherResult.Builder<Object> resultBuilder) throws Exception {
+            eventEmitter.fireBeforeDataFetch();
+
+            try {
+                Object[] transformedArguments = argumentHelper.getArguments(dfe);
+                Object resultFromMethodCall = reflectionHelper.invoke(transformedArguments);
+                Object resultFromTransform = fieldHelper.transformResponse(resultFromMethodCall);
+                resultBuilder.data(resultFromTransform);
+            } catch (AbstractDataFetcherException abstractDataFetcherException) {
+                //Arguments or result couldn't be transformed
+                abstractDataFetcherException.appendDataFetcherResult(resultBuilder, dfe);
+                eventEmitter.fireOnDataFetchError(dfe.getExecutionId().toString(), abstractDataFetcherException);
+            } catch (GraphQLException graphQLException) {
+                partialResultHelper.appendPartialResult(resultBuilder, dfe, graphQLException);
+                eventEmitter.fireOnDataFetchError(dfe.getExecutionId().toString(), graphQLException);
+            } catch (SecurityException | IllegalAccessException | IllegalArgumentException ex) {
+                //m.invoke failed
+                eventEmitter.fireOnDataFetchError(dfe.getExecutionId().toString(), ex);
+                throw ex;
+            } finally {
+                eventEmitter.fireAfterDataFetch();
+            }
+
+            return (T) resultBuilder.build();
+        }
+
+        @Override
+        public <T> CompletionStage<List<T>> batch(BatchLoaderEnvironment ble, List<Object> keys) {
+            Object[] arguments = batchLoaderHelper.getArguments(keys, ble);
+            final ClassLoader tccl = Thread.currentThread().getContextClassLoader();
+            return CompletableFuture.supplyAsync(() -> (List<T>) reflectionHelper.invokePrivileged(tccl, arguments));
```

Comment:
Don't use supplyAsync without an executor. It picks a thread from the fork/join pool, you don't have any control.

Reply:
That sounds like we might need some sort of config mechanism for users to be able to specify the executor service they want to use.

Reply:
I think we need to open a separate issue for this. (the current code is exactly the same as above.) 
I opened https://github.com/smallrye/smallrye-graphql/issues/496 for this, we need to change this as part of another PR.